node/cn/handler_msg_test.go

```diff
@@ -0,0 +1,121 @@
+func generateMsg(t *testing.T, msgCode uint64, data interface{}) p2p.Msg {
+	size, r, err := rlp.EncodeToReader(data)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return p2p.Msg{
+		Code:    msgCode,
+		Size:    uint32(size),
+		Payload: r,
+	}
+}
+
+// prepareTestHandleNewBlockMsg creates structs for TestHandleNewBlockMsg_ tests.
+func prepareTestHandleNewBlockMsg(t *testing.T, mockCtrl *gomock.Controller, blockNum int) (*types.Block, p2p.Msg, *MockPeer, *mocks2.MockProtocolManagerFetcher) {
+	mockPeer := NewMockPeer(mockCtrl)
+
+	newBlock := newBlock(blockNum)
+	newBlock.ReceivedFrom = mockPeer
+	msg := generateMsg(t, NewBlockMsg, newBlockData{Block: newBlock, TD: big.NewInt(int64(blockNum))})
+
+	mockPeer.EXPECT().AddToKnownBlocks(newBlock.Hash()).Times(1)
+	mockPeer.EXPECT().GetID().Return(nodeids[0].String()).AnyTimes()
+
+	mockFetcher := mocks2.NewMockProtocolManagerFetcher(mockCtrl)
+	mockFetcher.EXPECT().Enqueue(nodeids[0].String(), newBlock).Times(1)
+
+	return newBlock, msg, mockPeer, mockFetcher
+}
+
+func TestHandleNewBlockMsg_LargeLocalPeerBlockScore(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	_, msg, mockPeer, mockFetcher := prepareTestHandleNewBlockMsg(t, mockCtrl, blockNum1)
+
+	pm := &ProtocolManager{}
+	pm.fetcher = mockFetcher
+
+	mockPeer.EXPECT().Head().Return(hash1, big.NewInt(blockNum1+1)).AnyTimes()
+
+	assert.NoError(t, handleNewBlockMsg(pm, mockPeer, msg))
+}
+
+func TestHandleNewBlockMsg_SmallLocalPeerBlockScore_NoSynchronise(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	block, msg, mockPeer, mockFetcher := prepareTestHandleNewBlockMsg(t, mockCtrl, blockNum1)
+
+	pm := &ProtocolManager{}
+	pm.fetcher = mockFetcher
+
+	mockPeer.EXPECT().Head().Return(hash1, big.NewInt(blockNum1-2)).AnyTimes()
+	mockPeer.EXPECT().SetHead(block.ParentHash(), big.NewInt(blockNum1-1)).Times(1)
+
+	currBlock := newBlock(blockNum1 - 1)
+	mockBlockChain := mocks.NewMockBlockChain(mockCtrl)
+	mockBlockChain.EXPECT().CurrentBlock().Return(currBlock).Times(1)
+	mockBlockChain.EXPECT().GetTd(currBlock.Hash(), currBlock.NumberU64()).Return(big.NewInt(blockNum1)).Times(1)
+
+	pm.blockchain = mockBlockChain
+
+	assert.NoError(t, handleNewBlockMsg(pm, mockPeer, msg))
+}
+
+func TestHandleTxMsg(t *testing.T) {
+	pm := &ProtocolManager{}
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockPeer := NewMockPeer(mockCtrl)
+
+	txs := types.Transactions{tx1}
+	msg := generateMsg(t, TxMsg, txs)
+
+	// If pm.acceptTxs == 0, nothing happens.
+	{
+		assert.NoError(t, handleTxMsg(pm, mockPeer, msg))
+	}
+	// If pm.acceptTxs == 1, TxPool.HandleTxMsg is called.
+	{
+		atomic.StoreUint32(&pm.acceptTxs, 1)
+		mockTxPool := mocks.NewMockTxPool(mockCtrl)
+		mockTxPool.EXPECT().HandleTxMsg(gomock.Eq(txs)).AnyTimes()
+		pm.txpool = mockTxPool
+
+		mockPeer.EXPECT().AddToKnownTxs(txs[0].Hash()).Times(1)
+		assert.NoError(t, handleTxMsg(pm, mockPeer, msg))
```

Comment:
Minor comment:
How about adding a `tx == nil in txs` case?

Reply:
I will add more tests in other PRs. Thanks!